File: Scripts/run_cd8et_ilcco_coloc.R
suppressWarnings({
  library(data.table)
  library(coloc)
})
plink_bin <- "My_MR_Project/plink.exe"
if (!file.exists(plink_bin)) plink_bin <- "plink.exe"
bfile <- "My_MR_Project/Reference/g1000_eur"
mr_csv <- "My_MR_Project/CD8ET_TopN_MR_Results.csv"
eqtl_file <- "My_MR_Project/Exposure/cd8et_esnp_table.tsv.gz"
out_file <- "My_MR_Project/Outcome/28604730-GCST004748-EFO_0001071.h.tsv"
exp_dir <- "My_MR_Project/Exposure"
out_csv <- "D:/2026YJ/My_MR_Project/CD8ET_ILCCO_coloc_sensitivity.csv"
mr <- fread(mr_csv)
mr[, p_ivw := as.numeric(p_ivw)]
mr <- mr[p_ivw < 5e-8]
targets <- gsub("cd8et_", "", mr$gene)
out <- fread(out_file, sep = "\t")
out <- out[, .(hm_rsid, hm_chrom, hm_pos, hm_beta, standard_error, hm_effect_allele, hm_other_allele, hm_effect_allele_frequency)]
out[, hm_chrom := as.character(hm_chrom)]
out[, hm_pos := as.integer(hm_pos)]
res_all <- NULL
e_all <- fread(eqtl_file, sep = "\t", select = c("GENE","RSID","CHR","POS","SPEARMANS_RHO","P_VALUE","A2_FREQ_ONEK1K"))
for (g in targets) {
  eg <- e_all[GENE == g]
  if (nrow(eg) == 0) next
  uniq <- unique(eg$CHR)
  chr <- as.character(uniq[which.max(tabulate(match(eg$CHR, uniq)))])
  center <- as.integer(median(eg$POS))
  start <- center - 1000000L
  end <- center + 1000000L
  e_reg <- EG[CHR == chr & POS >= start & POS <= end]
  o_reg <- out[hm_chrom == chr & hm_pos >= start & hm_pos <= end]
  if (nrow(e_reg) > 0) {
    tmp_assoc <- tempfile(pattern = paste0("assoc_", g), fileext = ".txt")
    fwrite(e_reg[, .(SNP = RSID, P = P_VALUE)], tmp_assoc, sep = "\t")
    tmp_out <- tempfile(pattern = paste0("clump_", g))
    args <- c("--bfile", bfile, "--clump", tmp_assoc, "--clump-p1", "1", "--clump-p2", "1", "--clump-kb", "10000", "--clump-r2", "0.001", "--out", tmp_out)
    suppressWarnings(system2(plink_bin, args = args, stdout = TRUE, stderr = TRUE))
    clumped_file <- paste0(tmp_out, ".clumped")
    if (file.exists(clumped_file)) {
      cl <- fread(clumped_file, fill = TRUE)
      if ("SNP" %in% names(cl)) e_reg <- e_reg[RSID %in% unique(cl$SNP)]
    }
  }
  z_corr <- NA_real_
  peak_shared <- 0
  peak_dist <- -1L
  if (nrow(e_reg) > 0 & nrow(o_reg) > 0) {
    e_reg[, z_eqtl := qnorm(pmin(pmax(1 - P_VALUE/2, .Machine$double.eps), 1 - .Machine$double.eps)) * sign(SPEARMANS_RHO)]
    e_reg[, se_eqtl := abs(SPEARMANS_RHO / qnorm(pmin(pmax(1 - P_VALUE/2, .Machine$double.eps), 1 - .Machine$double.eps)))]
    o_reg[, z_ilcco := hm_beta / standard_error]
    setnames(e_reg, "RSID", "SNP")
    setnames(o_reg, "hm_rsid", "SNP")
    j <- merge(e_reg[, .(SNP, CHR, POS, SPEARMANS_RHO, se_eqtl, A2_FREQ_ONEK1K)], o_reg[, .(SNP, hm_beta, standard_error, hm_effect_allele_frequency, hm_pos)], by = "SNP")
    z_eq <- j$SPEARMANS_RHO / j$se_eqtl
    z_gw <- j$hm_beta / j$standard_error
    keep_z <- which(is.finite(z_eq) & is.finite(z_gw))
    if (length(keep_z) >= 3) {
      med_eq <- median(z_eq[keep_z]); mad_eq <- mad(z_eq[keep_z]); thr_eq <- med_eq + 8 * mad_eq
      med_gw <- median(z_gw[keep_z]); mad_gw <- mad(z_gw[keep_z]); thr_gw <- med_gw + 8 * mad_gw
      mask <- (abs(z_eq) <= abs(thr_eq)) & (abs(z_gw) <= abs(thr_gw))
      z_eq <- z_eq[mask]
      z_gw <- z_gw[mask]
    }
    z_corr <- if (nrow(j) >= 3) suppressWarnings(cor(z_eq, z_gw)) else NA_real_
    if (nrow(j) >= 5) {
      dfj <- as.data.frame(j)
      dfj$z_eq <- z_eq
      dfj$z_gw <- z_gw
      fit <- suppressWarnings(lm(z_gw ~ z_eq, data = dfj))
      cd <- suppressWarnings(cooks.distance(fit))
      thr <- 4 / nrow(dfj)
      keep <- which(is.finite(cd) & cd <= thr)
      if (length(keep) >= 3) {
        z_corr <- suppressWarnings(cor(dfj$z_eq[keep], dfj$z_gw[keep]))
      }
    }
    peak_e <- e_reg[which.max(abs(z_eqtl))]
    peak_o <- o_reg[which.max(abs(z_ilcco))]
    peak_shared <- as.integer(peak_e$SNP == peak_o$SNP)
    peak_dist <- abs(as.integer(peak_e$POS) - as.integer(peak_o$hm_pos))
    pp4 <- NA_real_
    if (nrow(j) >= 10) {
      d1 <- list(beta = j$SPEARMANS_RHO, varbeta = (j$se_eqtl)^2, type = "quant", MAF = j$A2_FREQ_ONEK1K)
      d2 <- list(beta = j$hm_beta, varbeta = (j$standard_error)^2, type = "quant", MAF = j$hm_effect_allele_frequency)
      csum <- suppressWarnings(coloc.abf(d1, d2)$summary)
      if (!is.null(csum) && "PP.H4.abf" %in% names(csum)) pp4 <- as.numeric(csum["PP.H4.abf"])
    }
  } else {
    exp_file <- file.path(exp_dir, paste0("EXPOSURE_cd8et_", g, "_mr.tsv.gz"))
    if (file.exists(exp_file)) {
      d <- fread(exp_file, sep = "\t")
      d[, SNP := as.character(SNP)]
      o <- out
      setnames(o, c("hm_rsid","hm_effect_allele","hm_other_allele","hm_beta","standard_error","hm_chrom","hm_pos"), c("SNP","ea_outcome","nea_outcome","by","se_by","CHR","POS"))
      m <- merge(d, o[, .(SNP, ea_outcome, nea_outcome, by, se_by, CHR, POS)], by = "SNP")
      if (nrow(m) >= 2) {
        idx <- which(m$effect_allele != m$ea_outcome)
        if (length(idx) > 0) m$by[idx] <- -m$by[idx]
        z_eqtl_ins <- m$beta / m$se
        z_ilcco_ins <- m$by / m$se_by
        z_corr <- suppressWarnings(cor(z_eqtl_ins, z_ilcco_ins))
        pe <- m[which.max(abs(z_eqtl_ins))]
        po <- m[which.max(abs(z_ilcco_ins))]
        peak_shared <- as.integer(pe$SNP == po$SNP)
        peak_dist <- as.integer(abs(as.integer(pe$POS) - as.integer(po$POS)))
        chr <- as.character(modes <- names(sort(table(m$CHR), decreasing = TRUE))[1])
        center <- as.integer(median(m$POS))
      }
    }
  }
  exp_file <- file.path(exp_dir, paste0("EXPOSURE_cd8et_", g, "_mr.tsv.gz"))
  sens_Q <- NA_real_
  sens_Q_p <- NA_real_
  eg_int <- NA_real_
  eg_int_se <- NA_real_
  eg_int_p <- NA_real_
  conc <- NA_real_
  if (file.exists(exp_file)) {
    d <- fread(exp_file, sep = "\t")
    d[, SNP := as.character(SNP)]
    o <- out
    setnames(o, c("hm_rsid","hm_effect_allele","hm_other_allele","hm_beta","standard_error"), c("SNP","ea_outcome","nea_outcome","by","se_by"))
    m <- merge(d, o[, .(SNP, ea_outcome, nea_outcome, by, se_by)], by = "SNP")
    if (nrow(m) >= 2) {
      idx <- which(m$effect_allele != m$ea_outcome)
      if (length(idx) > 0) {
        m$by[idx] <- -m$by[idx]
      }
      bx <- m$beta
      by <- m$by
      se_bx <- m$se
      se_by <- m$se_by
      ratio <- by / bx
      var_ratio <- (se_by^2 / (bx^2)) + ((by^2) * (se_bx^2) / (bx^4))
      w <- 1 / var_ratio
      ivw <- sum(w * ratio) / sum(w)
      q <- sum(w * (ratio - ivw)^2)
      sens_Q <- q
      sens_Q_p <- 1 - pchisq(q, df = nrow(m) - 1)
      fit <- suppressWarnings(lm(by ~ bx, weights = 1/(se_by^2)))
      co <- coef(fit)
      vc <- vcov(fit)
      eg_int <- co[1]
      eg_int_se <- sqrt(vc[1,1])
      eg_int_p <- 2 * pnorm(abs(eg_int/eg_int_se), lower.tail = FALSE)
      conc <- mean((bx * by) > 0)
    }
  }
  shared <- if (exists("j")) nrow(j) else if (exists("m")) nrow(m) else 0L
  res <- data.table(gene = g, chr = chr, center_pos = center, n_eqtl_region = if (exists("m")) nrow(m) else nrow(e_reg), n_ilcco_region = if (exists("m")) nrow(o_reg) else nrow(o_reg), n_shared_rsids = shared, z_corr = z_corr, peak_shared = peak_shared, peak_dist_bp = peak_dist, Q = sens_Q, Q_p = sens_Q_p, egger_intercept = eg_int, egger_intercept_se = eg_int_se, egger_intercept_p = eg_int_p, concordance = conc, pp4 = if (exists("pp4")) pp4 else NA_real_)
  res_all <- rbind(res_all, res, fill = TRUE)
}
if (is.null(res_all)) {
  schema <- data.table(gene=character(), chr=character(), center_pos=integer(), n_eqtl_region=integer(), n_ilcco_region=integer(), n_shared_rsids=integer(), z_corr=double(), peak_shared=integer(), peak_dist_bp=integer(), Q=double(), Q_p=double(), egger_intercept=double(), egger_intercept_se=double(), egger_intercept_p=double(), concordance=double(), pp4=double())
  fwrite(schema, out_csv)
} else {
  fwrite(res_all, out_csv)
}
writeLines(paste0("saved ", out_csv))

File: Scripts/run_cd8et_ilcco_coloc.py
import pandas as pd
import numpy as np
from pathlib import Path
from statistics import NormalDist

MR = Path("My_MR_Project/CD8ET_TopN_MR_Results.csv")
EQTL = Path("My_MR_Project/Exposure/cd8et_eqtl_table.tsv.gz")
OUTCOME = Path("My_MR_Project/Outcome/28604730-GCST004748-EFO_0001071.h.tsv")
EXP_DIR = Path("My_MR_Project/Exposure")
OUTCSV = Path("CD8ET_ILCCO_coloc_sensitivity.csv")

def mode_chr(x):
    vals, counts = np.unique(x, return_counts=True)
    return str(vals[np.argmax(counts)])

def read_eqtl_genes(genes):
    cols = ['GENE','RSID','CHR','POS','SPEARMANS_RHO','P_VALUE']
    buf = []
    print(f"Reading eQTL file for {len(genes)} genes...", flush=True)
    try:
        for i, chunk in enumerate(pd.read_csv(EQTL, sep='\t', compression='gzip', usecols=cols, chunksize=500000)):
            if i % 1 == 0:
                print(f"Processing chunk {i}, buf size: {len(buf)}", flush=True)
            # if i > 20: break # Debug limit
            g = chunk[chunk['GENE'].isin(genes)]
            if len(g):
                buf.append(g)
    except Exception as e:
        print(f"Error reading eQTL file: {e}", flush=True)
        return None
        
    if not buf:
        return pd.DataFrame(columns=cols)
    return pd.concat(buf, ignore_index=True)

def main():
    mr = pd.read_csv(MR)
    mr['p_ivw'] = pd.to_numeric(mr['p_ivw'], errors='coerce')
    mr = mr[mr['p_ivw'] < 5e-8]
    targets = [x.replace('cd8et_','') for x in mr['gene'].astype(str).tolist()]
    print('targets_count', len(targets), flush=True)
    
    # Read all eQTL data for targets at once
    e_all = read_eqtl_genes(set(targets))
    if e_all is None:
        print("No eQTL data found for targets (None)", flush=True)
        return
    print(f"e_all rows: {len(e_all)}", flush=True)
    if len(e_all) == 0:
        print("No eQTL data found for targets (0 rows)", flush=True)
        # still write empty schema
        out_path = Path("My_MR_Project/CD8ET_ILCCO_coloc_sensitivity.csv")
        cols = ['gene','chr','center_pos','n_eqtl_region','n_ilcco_region','n_shared_rsids','z_corr','peak_shared','peak_dist_bp','Q','Q_p','egger_intercept','egger_intercept_se','egger_intercept_p','concordance']
        pd.DataFrame(columns=cols).to_csv(out_path, index=False)
        print('saved empty:', out_path)
        return
    
    out = pd.read_csv(OUTCOME, sep='\t', low_memory=False)
    out = out.rename(columns={'hm_rsid':'SNP','hm_chrom':'CHR','hm_pos':'POS','hm_beta':'by','standard_error':'se_by','hm_effect_allele':'ea_outcome','hm_other_allele':'nea_outcome'})
    out['CHR'] = out['CHR'].astype(str)
    out['POS'] = pd.to_numeric(out['POS'], errors='coerce')
    out = out.dropna(subset=['POS'])
    out['POS'] = out['POS'].astype(int)
    
    rows = []
    for g in targets:
        print(f"Processing gene {g}...", flush=True)
        eg = e_all[e_all['GENE'] == g].copy()
        if len(eg)==0:
            exp_file = EXP_DIR / f"EXPOSURE_cd8et_{g}_mr.tsv.gz"
            if not exp_file.exists():
                print(f"No data for {g}", flush=True)
                continue
            d = pd.read_csv(exp_file, sep='\t', compression='gzip')
            d['SNP'] = d['SNP'].astype(str)
            m = pd.merge(d, out[['SNP','ea_outcome','nea_outcome','by','se_by','CHR','POS']], on='SNP', how='inner')
            if len(m) < 2:
                print(f"No overlap instruments for {g}", flush=True)
                continue
            idx = m['effect_allele'] != m['ea_outcome']
            if idx.any():
                m.loc[idx, 'by'] = -m.loc[idx, 'by']
            bx = m['beta'].astype(float).values
            by = m['by'].astype(float).values
            se_bx = m['se'].astype(float).values
            se_by = m['se_by'].astype(float).values
            ratio = by / bx
            var_ratio = (se_by**2 / (bx**2)) + ((by**2) * (se_bx**2) / (bx**4))
            w = 1 / var_ratio
            ivw = float(np.sum(w * ratio) / np.sum(w))
            Q = float(np.sum(w * (ratio - ivw)**2))
            X = np.vstack([np.ones_like(bx), bx]).T
            W = np.diag(1.0/(se_by**2))
            XtW = X.T @ W
            beta_hat = np.linalg.pinv(XtW @ X) @ XtW @ by
            eg_int = float(beta_hat[0])
            resid = by - X @ beta_hat
            sigma2 = float((resid.T @ W @ resid) / (len(bx) - 2))
            cov = np.linalg.pinv(XtW @ X) * sigma2
            eg_int_se = float(np.sqrt(cov[0,0]))
            eg_int_p = float(2 * (1 - NormalDist().cdf(abs(eg_int / eg_int_se))))
            conc = float(np.mean(bx * by > 0))
            z_eqtl_ins = bx / se_bx
            z_ilcco_ins = by / se_by
            z_corr = float(pd.Series(z_eqtl_ins).corr(pd.Series(z_ilcco_ins)))
            pe_idx = int(np.argmax(np.abs(z_eqtl_ins)))
            po_idx = int(np.argmax(np.abs(z_ilcco_ins)))
            peak_shared = int(str(m.iloc[pe_idx]['SNP']) == str(m.iloc[po_idx]['SNP']))
            try:
                peak_dist = int(abs(int(m.iloc[pe_idx]['POS']) - int(m.iloc[po_idx]['POS'])))
            except Exception:
                peak_dist = -1
            chr_ = str(m['CHR'].astype(str).mode().iloc[0])
            center = int(np.median(m['POS'].astype(int).values))
            rows.append({'gene': g, 'chr': chr_ , 'center_pos': center, 'n_eqtl_region': int(len(m)), 'n_ilcco_region': int(len(m)), 'n_shared_rsids': int(len(m)), 'z_corr': z_corr, 'peak_shared': peak_shared, 'peak_dist_bp': peak_dist, 'Q': Q, 'Q_p': np.nan, 'egger_intercept': eg_int, 'egger_intercept_se': eg_int_se, 'egger_intercept_p': eg_int_p, 'concordance': conc})
            continue
        # sanitize POS
        eg['POS'] = pd.to_numeric(eg['POS'], errors='coerce')
        eg = eg.dropna(subset=['POS'])
        eg['POS'] = eg['POS'].astype(int)
        if len(eg)==0:
            print(f"No valid POS for {g}", flush=True)
            continue
        chr_ = mode_chr(eg['CHR'].astype(str).values)
        center = int(np.median(eg['POS'].values))
        start = center - 1000000
        end = center + 1000000
        e_reg = eg[(eg['CHR'].astype(str)==chr_) & (eg['POS'].astype(int)>=start) & (eg['POS'].astype(int)<=end)].copy()
        if len(e_reg)==0:
            # fall back to instrument-level colocalization when region extraction empty
            e_reg = pd.DataFrame(columns=['RSID','POS','z_eqtl'])
        pvals = e_reg['P_VALUE'].astype(float).values
        pvals = np.clip(pvals, 1e-300, 1.0)
        z_eqtl = []
        rhos = e_reg['SPEARMANS_RHO'].values
        for i, p in enumerate(pvals):
            if np.isfinite(p) and 0.0 < float(1 - float(p)/2) < 1.0:
                z = NormalDist().inv_cdf(1 - float(p)/2)
            else:
                z = np.nan
            z_eqtl.append(z * np.sign(rhos[i]))
        e_reg['z_eqtl'] = np.array(z_eqtl)
        o_reg = out[(out['CHR']==chr_) & (out['POS']>=start) & (out['POS']<=end)].copy()
        if len(o_reg)==0:
            # fall back to instrument-level colocalization when region extraction empty
            o_reg = pd.DataFrame(columns=['SNP','POS','z_ilcco'])
        # compute z with guard against division by zero
        se_safe = o_reg['se_by'].replace(0, np.nan)
        o_reg['z_ilcco'] = o_reg['by'].values / se_safe.values
        j = pd.merge(e_reg[['RSID','z_eqtl','POS']].rename(columns={'RSID':'SNP'}), o_reg[['SNP','z_ilcco','POS']].rename(columns={'POS':'POS_OUT'}), on='SNP', how='inner')
        z_corr = float(j[['z_eqtl','z_ilcco']].corr().iloc[0,1]) if len(j)>=3 else np.nan
        
        # Filter out rows with NaN/inf z-scores before argmax
        e_reg_valid = e_reg[np.isfinite(e_reg['z_eqtl'])]
        o_reg_valid = o_reg[np.isfinite(o_reg['z_ilcco'])]
        
        if len(e_reg_valid) > 0 and len(o_reg_valid) > 0:
            pe = e_reg_valid.iloc[int(np.argmax(np.abs(e_reg_valid['z_eqtl'].values)))]
            po = o_reg_valid.iloc[int(np.argmax(np.abs(o_reg_valid['z_ilcco'].values)))]
            peak_shared = int(str(pe['RSID']) == str(po['SNP']))
            peak_dist = int(abs(int(pe['POS']) - int(po['POS'])))
        else:
            # instrument-level fallback if possible
            exp_file = EXP_DIR / f"EXPOSURE_cd8et_{g}_mr.tsv.gz"
            if exp_file.exists():
                d = pd.read_csv(exp_file, sep='\t', compression='gzip')
                d['SNP'] = d['SNP'].astype(str)
                d['z_eqtl_ins'] = d['beta'].astype(float) / d['se'].astype(float)
                mo = pd.merge(d[['SNP','z_eqtl_ins']], out[['SNP','by','se_by','POS']], on='SNP', how='inner')
                if len(mo) >= 2:
                    z_ilcco_ins = mo['by'].astype(float) / mo['se_by'].replace(0, np.nan).astype(float)
                    z_corr = float(pd.Series(mo['z_eqtl_ins']).corr(z_ilcco_ins))
                    pe_idx = int(np.argmax(np.abs(mo['z_eqtl_ins'].values)))
                    po_idx = int(np.argmax(np.abs(z_ilcco_ins.values)))
                    peak_shared = int(str(mo.iloc[pe_idx]['SNP']) == str(mo.iloc[po_idx]['SNP']))
                    # use outcome POS for distance when available
                    try:
                        peak_dist = int(abs(int(mo.iloc[pe_idx]['POS']) - int(mo.iloc[po_idx]['POS'])))
                    except Exception:
                        peak_dist = -1
                else:
                    peak_shared = 0
                    peak_dist = -1
            else:
                peak_shared = 0
                peak_dist = -1
        exp_file = EXP_DIR / f"EXPOSURE_cd8et_{g}_mr.tsv.gz"
        Q = np.nan; Q_p = np.nan; eg_int = np.nan; eg_int_se = np.nan; eg_int_p = np.nan; conc = np.nan
        if exp_file.exists():
            d = pd.read_csv(exp_file, sep='\t', compression='gzip')
            d['SNP'] = d['SNP'].astype(str)
            m = pd.merge(d, out[['SNP','ea_outcome','nea_outcome','by','se_by']], on='SNP', how='inner')
            if len(m)>=2:
                idx = m['effect_allele'] != m['ea_outcome']
                if idx.any():
                    m.loc[idx, 'by'] = -m.loc[idx, 'by']
                bx = m['beta'].astype(float).values
                by = m['by'].astype(float).values
                se_bx = m['se'].astype(float).values
                se_by = m['se_by'].astype(float).values
                ratio = by / bx
                var_ratio = (se_by**2 / (bx**2)) + ((by**2) * (se_bx**2) / (bx**4))
                w = 1 / var_ratio
                ivw = float(np.sum(w * ratio) / np.sum(w))
                Q = float(np.sum(w * (ratio - ivw)**2))
                Q_p = np.nan
                X = np.vstack([np.ones_like(bx), bx]).T
                W = np.diag(1.0/(se_by**2))
                XtW = X.T @ W
                beta_hat = np.linalg.pinv(XtW @ X) @ XtW @ by
                eg_int = float(beta_hat[0])
                resid = by - X @ beta_hat
                sigma2 = float((resid.T @ W @ resid) / (len(bx) - 2))
                cov = np.linalg.pinv(XtW @ X) * sigma2
                eg_int_se = float(np.sqrt(cov[0,0]))
                eg_int_p = float(2 * (1 - NormalDist().cdf(abs(eg_int / eg_int_se))))
                conc = float(np.mean(bx * by > 0))
        rows.append({'gene': g, 'chr': chr_, 'center_pos': center, 'n_eqtl_region': int(len(e_reg)), 'n_ilcco_region': int(len(o_reg)), 'n_shared_rsids': int(len(j)), 'z_corr': z_corr, 'peak_shared': peak_shared, 'peak_dist_bp': peak_dist, 'Q': Q, 'Q_p': Q_p, 'egger_intercept': eg_int, 'egger_intercept_se': eg_int_se, 'egger_intercept_p': eg_int_p, 'concordance': conc})
    print('rows_len', len(rows))
    if rows:
        out_path = Path("My_MR_Project/CD8ET_ILCCO_coloc_sensitivity.csv")
        pd.DataFrame(rows).to_csv(out_path, index=False)
        print('saved:', out_path)
    else:
        cols = ['gene','chr','center_pos','n_eqtl_region','n_ilcco_region','n_shared_rsids','z_corr','peak_shared','peak_dist_bp','Q','Q_p','egger_intercept','egger_intercept_se','egger_intercept_p','concordance']
        out_path = Path("My_MR_Project/CD8ET_ILCCO_coloc_sensitivity.csv")
        pd.DataFrame(columns=cols).to_csv(out_path, index=False)
        print('saved empty:', out_path)

if __name__ == '__main__':
    try:
        main()
    except Exception as e:
        Path("My_MR_Project/coloc_err.txt").write_text(str(e))
